algorithms/src/polycommit/sonic_pc/data_structures.rs

```diff
@@ -477,5 +488,279 @@ impl<E: PairingEngine> Prepare<PreparedVerifierKey<E>> for VerifierKey<E> {
+    pub fn push(&mut self, term: (F, LCTerm)) -> &mut Self {
+        self.terms.push(term);
+        self
+    }
+}
+
+impl<'a, F: Field> AddAssign<(F, &'a LinearCombination<F>)> for LinearCombination<F> {
+    #[allow(clippy::suspicious_op_assign_impl)]
+    fn add_assign(&mut self, (coeff, other): (F, &'a LinearCombination<F>)) {
+        self.terms.extend(other.terms.iter().map(|(c, t)| (coeff * c, t.clone())));
+    }
+}
+
+impl<'a, F: Field> SubAssign<(F, &'a LinearCombination<F>)> for LinearCombination<F> {
+    #[allow(clippy::suspicious_op_assign_impl)]
+    fn sub_assign(&mut self, (coeff, other): (F, &'a LinearCombination<F>)) {
+        self.terms.extend(other.terms.iter().map(|(c, t)| (-coeff * c, t.clone())));
+    }
+}
+
+impl<'a, F: Field> AddAssign<&'a LinearCombination<F>> for LinearCombination<F> {
+    fn add_assign(&mut self, other: &'a LinearCombination<F>) {
+        self.terms.extend(other.terms.iter().cloned());
+    }
+}
+
+impl<'a, F: Field> SubAssign<&'a LinearCombination<F>> for LinearCombination<F> {
+    fn sub_assign(&mut self, other: &'a LinearCombination<F>) {
+        self.terms.extend(other.terms.iter().map(|(c, t)| (-*c, t.clone())));
+    }
+}
+
+impl<F: Field> AddAssign<F> for LinearCombination<F> {
+    fn add_assign(&mut self, coeff: F) {
+        self.terms.push((coeff, LCTerm::One));
+    }
+}
+
+impl<F: Field> SubAssign<F> for LinearCombination<F> {
+    fn sub_assign(&mut self, coeff: F) {
+        self.terms.push((-coeff, LCTerm::One));
+    }
+}
+
+impl<F: Field> MulAssign<F> for LinearCombination<F> {
+    fn mul_assign(&mut self, coeff: F) {
+        self.terms.iter_mut().for_each(|(c, _)| *c *= &coeff);
+    }
+}
+
+impl<F: Field> core::ops::Deref for LinearCombination<F> {
+    type Target = [(F, LCTerm)];
+
+    fn deref(&self) -> &Self::Target {
+        &self.terms
+    }
+}
+
+/// `QuerySet` is the set of queries that are to be made to a set of labeled polynomials/equations
+/// `p` that have previously been committed to. Each element of a `QuerySet` is a `(label, query)`
+/// pair, where `label` is the label of a polynomial in `p`, and `query` is the field element
+/// that `p[label]` is to be queried at.
+///
+/// Added the third field: the point name.
+pub type QuerySet<'a, T> = BTreeSet<(String, (String, T))>;
+
+/// `Evaluations` is the result of querying a set of labeled polynomials or equations
+/// `p` at a `QuerySet` `Q`. It maps each element of `Q` to the resulting evaluation.
+/// That is, if `(label, query)` is an element of `Q`, then `evaluation.get((label, query))`
+/// should equal `p[label].evaluate(query)`.
+pub type Evaluations<'a, F> = BTreeMap<(String, F), F>;
+
+/// Evaluate the given polynomials at `query_set`.
+pub fn evaluate_query_set<'a, F: PrimeField>(
+    polys: impl IntoIterator<Item = &'a LabeledPolynomial<F>>,
+    query_set: &QuerySet<'a, F>,
+) -> Evaluations<'a, F> {
+    let polys: BTreeMap<_, _> = polys.into_iter().map(|p| (p.label(), p)).collect();
```

Comment:
does this have to be a `BTreeMap`? lookups are much faster in a `HashMap`

Reply:
Hm ordering is indeed not important here. Happy to change it to a HashMap, but generally the size of this Map is going to be quite small (20-30 items *at most*), so I don't think it will have a big perf impact.

Reply:
if it's that small, then a vector would probably be fine too